tests/manualGrading.test.js

```diff
@@ -0,0 +1,408 @@
+const { assert } = require('chai');
+const cheerio = require('cheerio');
+const fs = require('fs-extra');
+const path = require('path');
+
+const config = require('../lib/config');
+const fetch = require('node-fetch');
+const helperServer = require('./helperServer');
+const sqlLoader = require('../prairielib/lib/sql-loader');
+const sqlDb = require('../prairielib/lib/sql-db');
+const sql = sqlLoader.loadSqlEquiv(__filename);
+const { setUser, parseInstanceQuestionId, saveOrGrade } = require('./helperClient');
+
+const siteUrl = 'http://localhost:' + config.serverPort;
+const baseUrl = siteUrl + '/pl';
+const defaultUser = {
+  authUid: config.authUid,
+  authName: config.authName,
+  authUin: config.authUin,
+};
+
+const fibonacciSolution = fs.readFileSync(
+  path.resolve(
+    __dirname,
+    '..',
+    'testCourse',
+    'questions',
+    'externalGrade',
+    'codeUpload',
+    'tests',
+    'ans.py'
+  )
+);
+
+const mockStudents = [
+  { authUid: 'student1', authName: 'Student User 1', authUin: '00000001' },
+  { authUid: 'student2', authName: 'Student User 2', authUin: '00000002' },
+  { authUid: 'student3', authName: 'Student User 3', authUin: '00000003' },
+  { authUid: 'student4', authName: 'Student User 4', authUin: '00000004' },
+];
+
+const mockStaff = [
+  { authUid: 'staff1', authName: 'Staff User 1', authUin: 'STAFF001' },
+  { authUid: 'staff2', authName: 'Staff User 2', authUin: 'STAFF002' },
+  { authUid: 'staff3', authName: 'Staff User 3', authUin: 'STAFF003' },
+  { authUid: 'staff4', authName: 'Staff User 4', authUin: 'STAFF004' },
+];
+
+const assessmentTitle = 'Homework for Internal, External, Manual grading methods';
+const manualGradingQuestionTitle = 'Manual Grading: Fibonacci function, file upload';
+
+/**
+ * @param {object} student or instructor user to load page by
+ * @returns string Returns "Homework for Internal, External, Manual grading methods" page text
+ */
+const loadHomeworkPage = async (user) => {
+  setUser(user);
+  const studentCourseInstanceUrl = baseUrl + '/course_instance/1';
+  const courseInstanceBody = await (await fetch(studentCourseInstanceUrl)).text();
+  const $courseInstancePage = cheerio.load(courseInstanceBody);
+  const hm9InternalExternalManualUrl =
+    siteUrl + $courseInstancePage(`a:contains("${assessmentTitle}")`).attr('href');
+  let res = await fetch(hm9InternalExternalManualUrl);
+  assert.equal(res.ok, true);
+  return res.text();
+};
+
+const loadHomeworkQuestionUrl = async (user) => {
+  const hm1Body = await loadHomeworkPage(user);
+  const $hm1Body = cheerio.load(hm1Body);
+  return siteUrl + $hm1Body(`a:contains("${manualGradingQuestionTitle}")`).attr('href');
+};
+
+/**
+ * Gets the score text for the first submission panel on the page.
+ *
+ * @param {import('cheerio')} $
+ * @returns {string}
```

Comment:
I appreciate the effort to add type annotations! Could you add `// @ts-check` as the first line of this file? That will ensure that this file is actually type-checked when `make typecheck` is run.

Reply:
> but it will take time as that would require the entire code to be reviewed and type annotated

I'm not asking for everything to be type-annotated, just that we opt the file in to typechecking. I added the `// @ts-check` annotation locally and only two lines had to change for typechecking to succeed:

```diff
- const fetch = require('node-fetch');
+ const fetch = require('node-fetch').default;
...
-           submission_score_percent: score_percent,
+           submission_score_percent: score_percent.toString(),
```

Much of the code will still use implicit `any` for types, but a lot of it (especially third-party code like `node-fetch`, `fs-extra`, `path`, `chai`, `cheerio`, etc...) does have useful type coverage!